Add Controller factory

markstory opened the pull request:
Having this logic out of the a dispatcher filter is necessary for PSR7 migration where the controller factory acts as part of the `ActionDispatcher`

Refs #6960

markstory committed:
Add standalone ControllerFactory.

Having this logic out of the a dispatcher filter is necessary for PSR7
migration where the controller factory acts as part of the
ActionDispatcher.

markstory committed:
Remove duplicate code.

Re-use the standalone ControllerFactory in the dispatcher filter. This
allows the Dispatcher to be cleaned up a bit as exceptions for missing
controllers are raised earlier in the process.

This is a minor behavior change, but a reasonable one in my eyes.

jippi commented on src/Routing/Dispatcher.php:
```diff
@@ -70,19 +69,13 @@ public function dispatch(Request $request, Response $response)
-                'prefix' => empty($request->params['prefix']) ? null : $request->params['prefix'],
-                '_ext' => empty($request->params['_ext']) ? null : $request->params['_ext']
-            ]);
+        if (!isset($beforeEvent->data['controller'])) {
```
nice, so now any filter can return a controller? :D

lorenzo replied:
It has always been like that

lorenzo commented:
Tests are failing

markstory commented:
@lorenzo I'll fix the tests. I suspect I forgot to commit a file.

markstory committed:
Fix failing tests.

TestPluginThree is now autoloaded by composer. We can no longer use that
plugin to test autoloading registration. Switch the autoload tests to
use TestPluginFive which is not in the autoloader.

markstory commented:
Tests should be passing now.

markstory committed:
Disallow controller names with / in them.

Controller names should not be allowed to have / in them. Internally we
convert / into \\ which allows arbitrary namespace creation through what
should be controlled parameters. While the default routing normally
prevents / getting into a controller name, we cannot make the same
assumptions with PSR7 middleware.

lorenzo commented:
Good job!
